pkg/component/registrycaches/registry_caches.go

```diff
@@ -71,30 +82,94 @@ type Values struct {
 	// before ManagedResource deletion during the Destroy operation. When set to true, the deployed
 	// resources by ManagedResources won't be deleted, but the ManagedResource itself will be deleted.
 	KeepObjectsOnDestroy bool
+	// CacheStatuses are the registry cache statuses, if any.
+	CacheStatuses []api.RegistryCacheStatus
+	// RegistryStatus is an empty registry status to fill.
+	RegistryStatus *v1alpha3.RegistryStatus
 }
 
 // New creates a new instance of DeployWaiter for registry caches.
 func New(
 	client client.Client,
+	shootClient client.Client,
+	secretManager secretsmanager.Interface,
+	logger logr.Logger,
 	namespace string,
 	values Values,
 ) component.DeployWaiter {
 	return &registryCaches{
-		client:    client,
-		namespace: namespace,
-		values:    values,
+		client:        client,
+		shootClient:   shootClient,
+		secretManager: secretManager,
+		logger:        logger,
+		namespace:     namespace,
+		values:        values,
 	}
 }
 
 type registryCaches struct {
-	client    client.Client
-	namespace string
-	values    Values
+	client        client.Client
+	shootClient   client.Client
+	secretManager secretsmanager.Interface
+	logger        logr.Logger
+	namespace     string
+	values        Values
 }
 
 // Deploy implements component.DeployWaiter.
 func (r *registryCaches) Deploy(ctx context.Context) error {
-	data, err := r.computeResourcesData(ctx)
+	// TODO(dimitar-kostadinov): If services are previously created with ManagedResource remove service object references from ManagedResource status - remove this after v0.11.0
+	mr := &resourcesv1alpha1.ManagedResource{
+		ObjectMeta: metav1.ObjectMeta{Name: managedResourceName, Namespace: r.namespace},
+	}
+	err := r.client.Get(ctx, client.ObjectKeyFromObject(mr), mr)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+	var updatedRefs []resourcesv1alpha1.ObjectReference
+	for _, objectRef := range mr.Status.Resources {
+		if objectRef.Kind != "Service" {
+			updatedRefs = append(updatedRefs, objectRef)
+		}
+	}
+	if len(updatedRefs) != len(mr.Status.Resources) {
+		patch := client.MergeFrom(mr.DeepCopy())
+		mr.Status.Resources = updatedRefs
+		err = r.client.Status().Patch(ctx, mr, patch)
+		if err != nil {
+			return fmt.Errorf("failed to update ManagedResource status: %w", err)
+		}
+	}
+
+	//create registry cache services
+	if err := r.createServices(ctx); err != nil {
```

Comment:
In https://github.com/ialidzhikov/gardener-extension-registry-cache/tree/poc-1 I did a PoC for deploying the Services with a ManagedResource. The PoC is uses 2 ManagedResources, 1 component. I liked this approach more than the current approach with client. A lot of boilerplate was removed.
Next, I want to try out locally 2 components, 2 ManagedResources.

Reply:
> Next, I want to try out locally 2 components, 2 ManagedResources.

I did the PoC in https://github.com/ialidzhikov/gardener-extension-registry-cache/tree/poc-2. I rather vote for this option. The components are easier to tests. We are doing less in the components. The registryCaches component no longer fills the registry cache status for later use from the actuator.
But let me know what you think.

Reply:
I also prefer the option with 2 component and 2 MRs.